hvsock_test.go

```diff
@@ -0,0 +1,627 @@
+	// if it was sent before, the read would succeed if the data was buffered prior
+	_, err = cl.Read(b)
+	u.Is(err, windows.WSAESHUTDOWN, "client did not shutdown reads")
+
+	// safe to call multiple times
+	u.Must(cl.CloseRead(), "client second close write")
+
+	l.Close()
+	cl.Close()
+
+	wantErr := socket.ErrSocketClosed
+	u.Is(cl.CloseWrite(), wantErr, "client close write")
+	u.Is(cl.CloseRead(), wantErr, "client close read")
+}
+
+func TestHvSockDialNoTimeout(t *testing.T) {
+	u := newUtil(t)
+	ch := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		defer close(ch)
+
+		addr := randHvsockAddr()
+		cl, err := Dial(ctx, addr)
+		if err == nil {
+			cl.Close()
+		}
+		u.Is(err, windows.WSAECONNREFUSED)
+	}()
+
+	// connections usually take about ~500µs
+	u.Wait(ch, 2*time.Millisecond, "dial did not time out")
+}
+
+func TestHvSockDialDeadline(t *testing.T) {
+	u := newUtil(t)
+	d := &HvsockDialer{}
+	d.Deadline = time.Now().Add(50 * time.Microsecond)
+	d.Retries = 1
+	// we need the wait time to be long enough for the deadline goroutine to run first and signal
+	// timeout
+	d.RetryWait = 100 * time.Millisecond
+	addr := randHvsockAddr()
+	cl, err := d.Dial(context.Background(), addr)
+	if err == nil {
+		cl.Close()
+		t.Fatalf("dial should not have finished")
+	}
+	u.Is(err, context.DeadlineExceeded, "dial did not exceed deadline")
+}
+
+func TestHvSockDialContext(t *testing.T) {
+	u := newUtil(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	time.AfterFunc(50*time.Microsecond, cancel)
+
+	d := &HvsockDialer{}
+	d.Retries = 1
+	d.RetryWait = 100 * time.Millisecond
+	addr := randHvsockAddr()
+	cl, err := d.Dial(ctx, addr)
+	if err == nil {
+		cl.Close()
+		t.Fatalf("dial should not have finished")
+	}
+	u.Is(err, context.Canceled, "dial was not canceled")
+}
+
+func TestHvSockAcceptClose(t *testing.T) {
+	u := newUtil(t)
+	l, _ := serverListen(u)
+	go func() {
+		time.Sleep(50 * time.Millisecond)
+		l.Close()
+	}()
+
+	c, err := l.Accept()
+	if err == nil {
+		c.Close()
+		t.Fatal("listener should not have accepted anything")
+	}
+	u.Is(err, ErrFileClosed)
+}
+
+func FuzzRxTx(f *testing.F) {
+	for _, b := range [][]byte{
+		[]byte("hello?"),
+		[]byte("This is a really long string that should be a good example of the really long " +
+			"payloads that may be sent over hvsockets when really long inputs are being used, tautologically. " +
+			"That means that we will have to test with really long input sequences, which means that " +
+			"we need to include really long byte sequences or strings in our testing so that we know that " +
+			"the sockets can deal with really long inputs. Look at this key mashing: " +
+			"sdflhsdfgkjdhskljjsad;kljfasd;lfkjsadl ;fasdjfopiwej09q34iur092\"i4o[piwajfliasdkf-012ior]-" +
+			"01oi3;'lSD<Fplkasdjgoisaefjoiasdlj\"hgfoaisdkf';laksdjdf[poaiseefk-0923i4roi3qwjrf9" +
+			"08sEJKEFOLIsaejf[09saEJFLKSADjf;lkasdjf;kljaslddhgaskghk"),
```

Comment:
Commit of the year nominee

Reply:
Not to take away from the incredible sequence of characters here, but you can also make a multiline string with backquotes

```go

[]byte(`This is a really long string that should be a good example of the really long 
			payloads that may be sent over hvsockets when really long inputs are being used, tautologically. 
			That means that we will have to test with really long input sequences, which means that 
			we need to include really long byte sequences or strings in our testing so that we know that 
			the sockets can deal with really long inputs. Look at this key mashing: 
			sdflhsdfgkjdhskljjsad;kljfasd;lfkjsadl ;fasdjfopiwej09q34iur092\"i4o[piwajfliasdkf-012ior]-
			01oi3;'lSD<Fplkasdjgoisaefjoiasdlj\"hgfoaisdkf';laksdjdf[poaiseefk-0923i4roi3qwjrf9
			08sEJKEFOLIsaejf[09saEJFLKSADjf;lkasdjf;kljaslddhgaskghk`),
```

Reply:
wait, raw strings include the newlines and I cant escape things (though I doubt thats an issue)